docs/security/FreelistProtection.md

```diff
@@ -0,0 +1,127 @@
+# Protecting meta-data
+
+Corrupting an allocator's meta-data is a common pattern for increasing the power of a use-after-free or out-of-bounds write vulnerabilities.
+If you can corrupt the allocator's meta-data, then you can take a control gadget in one part of a system, and use it to affect other parts of the system.
+There are various approaches to protecting allocator meta-data, the most common are:
+
+* make the allocator meta-data hard to find through randomisation
+* surround meta-data with guard pages
+* add some level of encryption/checksuming
+
+With the refactoring of the page table ([described earlier](./VariableSizedChunks.md)), we can put all the slab meta-data at hard to find places with randomisation and add guard pages around it.
+But there is one super important bit that still remains: free lists.
+
+##  What are free lists?
+
+Many allocators chain together unused allocations into a linked list.
+This is super efficient as it doesn't require meta-data proportional to the number of allocations on a slab.
+The disused objects can be used in either a linked stack or queue.
+However, the key problem is neither randomisation or guard pages can be used to protect this data.
+
+In snmalloc, we are using a new technique for protecting this data.
+
+## Protecting a free queue.
+
+The idea is remarkably simple: a doubly linked list is far harder to corrupt than a single linked list, because you can check its invariant:
+```
+   x.next.prev == x
+```
+In every kind of free list in snmalloc, we now use something a kin to this:
+```
+  f(x.next).prev == g(x, f(x.next))
```

Comment:
This looks like you are `f(x.next)` as structure pointer but below it says that `f` is XOR with random values?!